File: scala-source/com/barrybecker4/puzzle/redpuzzle/solver/GeneticSearchSolver.scala
// Copyright by Barry G. Becker, 2017 - 2018. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.redpuzzle.solver

import com.barrybecker4.optimization.OptimizationListener
import com.barrybecker4.optimization.Optimizer
import com.barrybecker4.optimization.optimizee.Optimizee
import com.barrybecker4.optimization.parameter.{ParameterArray, ParameterArrayWithFitness}
import com.barrybecker4.puzzle.common.PuzzleController
import com.barrybecker4.puzzle.redpuzzle.model.{OrientedPiece, PieceList, PieceParameterArray}

import scala.collection.Seq
import com.barrybecker4.puzzle.redpuzzle.solver.FitnessFinder.MAX_FITS

import scala.util.Random

/**
  * Solve the red puzzle using a genetic search algorithm.
  * Solves the puzzle in 3.5 seconds on Core2 duo system (6 generations).
  * @author Barry Becker
  */
class GeneticSearchSolver(override val puzzle: PuzzleController[PieceList, OrientedPiece], val useConcurrency: Boolean)
                   extends RedPuzzleSolver(puzzle) with Optimizee with OptimizationListener {

  private val strategy =
    if (useConcurrency) com.barrybecker4.optimization.strategy.CONCURRENT_GENETIC_SEARCH
    else com.barrybecker4.optimization.strategy.GENETIC_SEARCH
  private val fitnessFinder = new FitnessFinder
  private var currentBestFitness = 10 + MAX_FITS

  /** @return list of moves to a solution. */
  def solve: Option[Seq[OrientedPiece]] = {
    val initialGuess = new PieceParameterArray(pieces, new Random(1))
    val startTime = System.currentTimeMillis
    val optimizer = new Optimizer(this)
    optimizer.setListener(this)
    val theSolution = optimizer.doOptimization(strategy, initialGuess, MAX_FITS)
    solution = theSolution.pa.asInstanceOf[PieceParameterArray].getPieceList
    val moves = if (evaluateFitness(theSolution.pa) == 0) Some(solution.pieces) else Option.empty
    val elapsedTime = System.currentTimeMillis - startTime
    puzzle.finalRefresh(moves, Option.apply(solution), numTries, elapsedTime)
    moves
  }

  def getName = "Genetic Search Solver for Red Puzzle"

  /** terminate the solver if we find a solution with this fitness. */
  def getOptimalFitness = 0

  def evaluateByComparison = false

  /** Return a low score if there are a lot of fits among the pieces.
    * @param params parameters
    * @return fitness value. Low is good.
    */
  def evaluateFitness(params: ParameterArray): Double = {
    val pieces = params.asInstanceOf[PieceParameterArray].getPieceList
    val fitness = fitnessFinder.calculateFitness(pieces)
    println("fitness = " + fitness)
    if (fitness < currentBestFitness) currentBestFitness = fitness
    fitness
  }

  def compareFitness(params1: ParameterArray, params2: ParameterArray): Double =
    throw new UnsupportedOperationException("compareFitness not used since we evaluate in an absolute way.")

  /** Called when the optimizer has made some progress optimizing.
    * Shows the current status. Update our current best guess at the solution.
    * @param params optimization parameters
    */
  def optimizerChanged(params: ParameterArrayWithFitness) {
    solution = params.pa.asInstanceOf[PieceParameterArray].getPieceList
    numTries += 1
    puzzle.refresh(solution, numTries)
    //System.out.println("current best = " + currentBestFitness)
  }
}

File: source/com/barrybecker4/puzzle/redpuzzle1/solver/GeneticSearchSolver.java
// Copyright by Barry G. Becker, 2000-2014. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.redpuzzle1.solver;

import com.barrybecker4.optimization.OptimizationListener;
import com.barrybecker4.optimization.Optimizer;
import com.barrybecker4.optimization.optimizee.Optimizee;
import com.barrybecker4.optimization.parameter.ParameterArray;
import com.barrybecker4.optimization.strategy.OptimizationStrategyType;
import com.barrybecker4.puzzle.common1.PuzzleController;
import com.barrybecker4.puzzle.redpuzzle1.model.Piece;
import com.barrybecker4.puzzle.redpuzzle1.model.PieceList;
import com.barrybecker4.puzzle.redpuzzle1.model.PieceParameterArray;
import scala.Option;
import scala.collection.JavaConversions;
import scala.collection.Seq;

import static com.barrybecker4.puzzle.redpuzzle1.solver.FitnessFinder.MAX_FITS;

/**
 * Solve the red puzzle using a genetic search algorithm.
 * Solves the puzzle in 3.5 seconds on Core2 duo system (6 generations).
 *
 * @author Barry Becker
 */
public class GeneticSearchSolver extends RedPuzzleSolver
                                 implements Optimizee, OptimizationListener {

    /** either genetic or concurrent genetic strategy. */
    private OptimizationStrategyType strategy;

    private FitnessFinder fitnessFinder;

    private double currentBestFitness = MAX_FITS;

    /** Constructor */
    GeneticSearchSolver(PuzzleController<PieceList, Piece> puzzle,
                               boolean useConcurrency) {
        super(puzzle);
        strategy = useConcurrency ? OptimizationStrategyType.CONCURRENT_GENETIC_SEARCH :
                                    OptimizationStrategyType.GENETIC_SEARCH;
        fitnessFinder = new FitnessFinder();
    }

    /**
     * @return list of moves to a solution.
     */
    @Override
    public Option<Seq<Piece>> solve()  {

        ParameterArray initialGuess = new PieceParameterArray(pieces_);
        solution_ = pieces_;
        long startTime = System.currentTimeMillis();

        Optimizer optimizer = new Optimizer(this);
        optimizer.setListener(this);

        ParameterArray solution =
            optimizer.doOptimization(strategy, initialGuess, MAX_FITS);

        solution_ = ((PieceParameterArray)solution).getPieceList();
        System.out.println("Solution = " + solution_);

        Option<Seq<Piece>> moves = Option.empty();
        if (evaluateFitness(solution) >= MAX_FITS) {
            moves = Option.apply(JavaConversions.asScalaBuffer(solution_.getPieces()).toSeq());
        }
        long elapsedTime = System.currentTimeMillis() - startTime;
        puzzle.finalRefresh(moves, Option.apply(solution_), numTries_, elapsedTime);

        return moves;
    }

    public String getName() {
         return "Genetic Search Solver for Red Puzzle";
    }

    /**
     * terminate the solver if we find a solution with this fitness.
     */
    public double getOptimalFitness() {
        return 0;
    }

    public boolean evaluateByComparison() {
        return false;
    }

    /**
     * Return a low score if there are a lot of fits among the pieces.
     *
     * @param params  parameters
     * @return fitness value. Low is good.
     */
    public double evaluateFitness(ParameterArray params) {
        PieceList pieces = ((PieceParameterArray) params).getPieceList();
        double fitness = fitnessFinder.calculateFitness(pieces);
        if (fitness < currentBestFitness) {
            currentBestFitness = fitness;
        }
        params.setFitness(fitness);
        return fitness;
    }

    public double compareFitness(ParameterArray params1, ParameterArray params2) {
        assert false : "compareFitness not used since we evaluate in an absolute way.";
        return 0;
    }

    /**
     * Called when the optimizer has made some progress optimizing.
     * Shows the current status.
     * @param params optimization parameters
     */
    public void optimizerChanged(ParameterArray params) {
        // update our current best guess at the solution.
        solution_ = ((PieceParameterArray) params).getPieceList();
        numTries_ ++;
        puzzle.refresh(solution_, numTries_);
        System.out.println("current best = " + currentBestFitness);
    }
}
